mlos_bench/mlos_bench/environments/script_env.py

```diff
@@ -0,0 +1,91 @@
+
+    def __init__(self,
+                 *,
+                 name: str,
+                 config: dict,
+                 global_config: Optional[dict] = None,
+                 tunables: Optional[TunableGroups] = None,
+                 service: Optional[Service] = None):
+        """
+        Create a new environment for script execution.
+
+        Parameters
+        ----------
+        name: str
+            Human-readable name of the environment.
+        config : dict
+            Free-format dictionary that contains the benchmark environment
+            configuration. Each config must have at least the "tunable_params"
+            and the "const_args" sections. It must also have at least one of
+            the following parameters: {setup, run, teardown}
+        global_config : dict
+            Free-format dictionary of global parameters (e.g., security credentials)
+            to be mixed in into the "const_args" section of the local config.
+        tunables : TunableGroups
+            A collection of tunable parameters for *all* environments.
+        service: Service
+            An optional service object (e.g., providing methods to
+            deploy or reboot a VM, etc.).
+        """
+        super().__init__(name=name, config=config, global_config=global_config,
+                         tunables=tunables, service=service)
+
+        self._script_setup = self.config.get("setup")
+        self._script_run = self.config.get("run")
+        self._script_teardown = self.config.get("teardown")
+
+        self._script_params: Optional[Iterable[str]] = self.config.get("script_params")
+        self._script_params_rename: Dict[str, str] = self.config.get("script_params_rename", {})
+
+        if self._script_setup is None and \
+           self._script_run is None and \
+           self._script_teardown is None:
+            raise ValueError("At least one of {setup, run, teardown} must be present")
+
+    def _get_env_params(self) -> Dict[str, str]:
+        """
+        Get the *shell* environment parameters to be passed to the script.
+
+        Returns
+        -------
+        env_params : Dict[str, str]
+            Parameters to pass as *shell* environment variables into the script.
+            This is usually a subset of `_params` with some possible conversions.
+        """
+        rename: Dict[str, str]
+        if self._script_params is None:
+            if self._script_params_rename:
+                # Only rename specified - use it.
+                rename = self._script_params_rename.copy()
+            else:
+                # Neither `script_params` nor rename are specified - use all params.
+                rename = {key: self._RE_INVALID.sub("_", key) for key in self._params}
+        else:
+            # Use `script_params` and rename if specified.
+            rename = {key: self._RE_INVALID.sub("_", key) for key in self._script_params}
+            rename.update(self._script_params_rename)
```

Comment:
I think listing a param in both `shell_params` and `shell_params_rename` is a config mistake, no?
Maybe prevent that?

Reply:
One other feature I could see being useful: regex based param names.
E.g. expose all params that match `^MLOS_[a-zA-Z0-9_]+$`
Could be a reasonable way to shorthand not explictly listing all args/settings names.
In that case the order of preference should probably go:
- regex
- explicit list overrides regex matches
- param map overrides regex matches
- any overlap between explicit list and param map are an error

Reply:
I left some comments about this in the schema files here: https://github.com/motus/MLOS/pull/5

Reply:
I think it is ok to have the same parameter in both `shell_params` and `shell_params_rename`. this way, if the user wants to keep the registry of all parameters that we pass to the script environment, we give them this option. I am also hesitant about the sophisticated renaming rules. I don't think the user should pass too many parameters to the environment anyway, so just having a sane default rule + explicit manual override should be sufficient

Reply:
On the later point I am fine.
In general though, I am worried about config behavior that requires reading the code to understand and prefer to give early feedback of potential issues as often as possible.
Anyways, it's fine for now.